### DISCUSSION TITLE
Integrated Terminal in Docker Desktop

### DISCUSSION THREAD

**Discussion by kimjoergensen** [docker/extension-ideas#127] [Ideas]
Currently when working with Docker Desktop you need an external terminal to run docker commands. It would be lovely to have an integrated terminal with the Docker Desktop application.

**Comment by amyb12345** [comment]
Hi! Thank you so much for the idea!  Have you seen the new integrated terminal in Docker Desktop that came out in release 4.12 ([https://docs.docker.com/desktop/use-desktop/container/](https://docs.docker.com/desktop/use-desktop/container/))?  If you have and it doesn't meet your needs, would you be willing to add more details on what else is needed? It would be really helpful to know for someone interested in building this experience!  Thanks!

**Reply by kimjoergensen** [reply]
Hello @amyb12345, I did notice the new integrated terminal from version 4.12, but to my best understanding it's only working with the container's STDIN - please let me know if that's wrong. What I am more interrested in, is having an integrated terminal that I can use for docker commands. It would be amazing if the terminal can open from the root of containers, so I can easily run docker commands from Docker files or docker-compose files.

My current workflow is to open the container from Visual Studio Code and using the integrated terminal there.

**Comment by kc-dot-io** [comment]
@kimjoergensen me too! so I built this [extension that loads a really nice ephemeral zsh terminal right into docker desktop](https://github.com/cto-ai/docker-extension-zsh) - still waiting to hear back on my extension marketplace submission but it can easily be installed from source!

**Reply by kimjoergensen** [reply]
This looks really nice, well done!

If you can get it to open with the directory from a container, then that's exactly what I am looking for!

**Reply by kc-dot-io** [reply]
So sort of like mounting a host directory or..?

**Reply by kimjoergensen** [reply]
Yes, mount a host directory from where the Dockerfile or docker-compose.yml file exists on the local drive.

**Reply by kc-dot-io** [reply]
@kimjoergensen yes great idea - I want this too. I was thinking about implementing but I wasn't quite sure how to infer it. If this were CLI, we could just use the cwd but since it's via Desktop that won't really work. What if we simply mounted $HOME into `/home/me` or something like this? This would also help setup another iternation that could benefit from session persistence inside of the desktop UI as well. Thoughts?

Let me know and I'll tag you in the PR when it's ready.
